src/riscv/lib/src/jit/builder/control_flow_graph.rs

```diff
@@ -0,0 +1,324 @@
+        let step_deltas = self.find_step_deltas();
+
+        // We pre-allocate space for the updates somewhat optimistically to avoid frequent
+        // re-allocations at the cost of some potential overallocation.
+        let mut updates = Vec::with_capacity(self.nodes.len().saturating_mul(2));
+
+        for (idx, node) in self.nodes.iter() {
+            // First, let's consider all forward edges to other nodes in the graph context. Those
+            // are eligible to update the step counter.
+            for edge in node.forward_edges.iter() {
+                let source_delta = step_deltas[idx];
+                let target_delta = step_deltas[edge.target];
+
+                // When the target delta is less or equal, that means the step counter is expected
+                // to be updated through edges to the target.
+                // The equal case is important for loops of instructions onto themselves. Those are
+                // step delta 0 -> 0 transitions.
+                if target_delta <= source_delta {
+                    updates.push(StepCounterUpdate {
+                        base_diff: source_delta - target_delta,
+                        edge: edge.info,
+                    });
+                }
+            }
+
+            // We mustn't forget exit edges as those are important points for updating the step
+            // counter as well. Forgetting these would mean that exiting a sequence of instructions
+            // would not update the step counter correctly.
+            for exit_edge in self.nodes[idx].exit_edges.iter() {
+                updates.push(StepCounterUpdate {
+                    base_diff: step_deltas[idx],
+                    edge: exit_edge,
+                });
+            }
+        }
+
+        updates.into_boxed_slice()
+    }
+}
+
+/// Directed edge where the target has been resolved to a node within the graph context
+type ResolvedDirectedEdge<'info, T> = self::DirectedEdgeInfo<&'info DirectedEdgeInfo<T>, InstrId>;
+
+/// Node in the control flow graph
+///
+/// This is usually information relating to an instruction.
+struct Node<'info, T> {
+    /// Edges to this node that originate from another node in this graph context
+    backward_edges: Box<[ResolvedDirectedEdge<'info, T>]>,
+
+    /// Edges from this node that target another node in the graph context
+    forward_edges: Box<[ResolvedDirectedEdge<'info, T>]>,
+
+    /// Exiting edges from this node that target outside of the graph context
+    exit_edges: Box<[&'info DirectedEdgeInfo<T>]>,
+
+    /// Information about this node
+    info: &'info NodeInfo<T>,
+}
+
+/// Classification of edges from a node
+enum EdgeClass<'edge, 'info, E> {
+    /// No attached edges (explicit or implicit)
+    Orphan,
+
+    // Exactly one explicit edge
+    SingleEdge(&'edge ResolvedDirectedEdge<'info, E>),
+
+    // Multiple explicit edges, or one explicit edge and one implicit edge (entrypoint)
+    MultipleEdges,
+}
```

Comment:
I think the notion of `SingleImplicit` is missing from these tests? It ends up mostly falling under `Orphan` - and while this works (step delta on an entrypoint with no explicit incomings is 0), it's not completely correct - and indeed commentary about what are (currently) Orphans not being reachable is incorrect.